### ISSUE TITLE
New liveness probes on consul-dataplane container prevents pods created by Kubernetes jobs to terminate

### ISSUE THREAD

**Issue by reneeckstein** [hashicorp/consul-k8s#1791]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

---

### Overview of the Issue

Until Consul v1.13.2 it was possible to stop the container `envoy-sidecar` by sending a POST request to `http://localhost:19000/quitquitquit` from the main container. This was very useful for Kubernetes Jobs to stop the sidecar after the job is completed successfully.

After upgrading to Consul v1.14.2 it still works to quit Envoy in the now-called `consul-dataplane` container with this POST request, but due to the newly added liveness probe, the container gets immediately restarted which prevents the pod to terminate successfully. The pod just keeps running in the "NotReady" state forever.

I didn't find any documentation about the added liveness probe in the Release Notes of either Consul or consul-k8s as already mentioned here -> https://github.com/hashicorp/consul-k8s/pull/1482#issuecomment-1330862189

While I agree this is very useful for usual pods, it is not for pods created by Kubernetes jobs. Maybe the liveness probes can be turned on/off via a new annotation or in general can be configured somehow.

### Reproduction Steps

Create a Connect-Inject-enabled Kubernetes job on Consul 1.14.x and add something like this to the end of your entrypoint script.

```
if [ "${USE_CONSUL_CONNECT}" == "true" ]
then
  echo "send shutdown signal to Consul sidecar"
  wget -O /dev/null --post-data -S 'http://localhost:19000/quitquitquit'
fi
```

### Expected behavior

Kubernetes jobs are successfully completed after sending the POST request to the `/quitquitquit` endpoint, consul-dataplane container is not restarted automatically.

### Environment details

Consul 1.14.2, Consul Helm chart 1.0.2 on EKS 1.22

**Comment by david-yu** [comment]
Hi @reneeckstein We do need to address this and have started to design a solution to address sidecar proxy lifecycle holistically inclusive of jobs. This is also a duplicate of https://github.com/hashicorp/consul-k8s/issues/536. We are hoping to start the work on this early next year.

**Comment by narendrapatel** [comment]
Hi,
I think this is more to do with the exit code from consul data plane rather then the health checks.
The following code in consul [dataplane](https://github.com/hashicorp/consul-dataplane/blob/main/pkg/consuldp/consul_dataplane.go) seems to be the problem:
```
case <-proxy.Exited():
	doneCh <- errors.New("envoy proxy exited unexpectedly")
```
I patched it to rather return 0 exit code and build and deployed a new image and the k8s jobs got completed successfully for me after that.
```
case <-proxy.Exited():
        cdp.logger.Info("envoy proxy exited")
        doneCh <- nil
```
But that is the hot fix and not a proper solution. A proper fix is to emit the exact exit code as is returned by envoy.
I'm not very familiar with the data plane code base and go, so unsure of how to achieve that.

**Comment by vyanamandra** [comment]
tl;dr --> sending SIGTERM to consul-dataplane helped me.


When I added `shareProcessNamespace: true` to the `job.spec.template.spec`, plus - added a command to the effect of - 
`kill -TERM $(ps -ef | grep '/usr/local/bin/consul-dataplane' | egrep -v 'grep|dumb-init' | awk '{print $2}')` after my main application ran, it did complete my entire pod.

This fix may not work as intended if an app-container user cannot send a SIGTERM to a PID due to a lack of permissions.

**Comment by curtbushko** [comment]
As part of consul-k8s releases, [1.0.8](https://github.com/hashicorp/consul-k8s/releases/tag/v1.0.8), [1.1.3](https://github.com/hashicorp/consul-k8s/releases/tag/v1.1.3) we added support for graceful shutdown in the proxy lifecycle. The next 1.2.x release that comes out will also have this feature enabled.

With it the feature, you can call `/graceful_shutdown` on the proxy and it will terminate consul-dataplane.

An example of how to use `/graceful_shutdown` can be seen in the [jobs example](https://developer.hashicorp.com/consul/docs/k8s/connect#kubernetes-jobs) on our website.

The most important piece is that you can curl the endpoint:
```
curl --max-time 2 -s -f -XPOST http://127.0.0.1:20600/graceful_shutdown
```

**Comment by david-yu** [comment]
Hi @reneeckstein will go ahead and close the issue as we hope this solves your issue. Please feel free to open a new issue if you have any new issues on the new feature we just delivered.

**Comment by david-yu** [comment]
@reneeckstein this should arrive within mid-august timeframe. We are aware it did not make it in for 1.2.1 as we wanted to restrict risky features from impacting our 1.16 GA. Thanks for your patience.
